### PR TITLE
Detemplate distribute sparsity pattern

### PR DESCRIPTION
**Pull request by drwells**
As was noted in the discussion in #1941: these functions are currently templates on the sparsity pattern type, but are only instantiated (in the `.cc` file) for exactly one sparsity pattern type each. Therefore they do not really need to be template functions.

### CODE REVIEW

**Comment by drwells** [comment]
This caused a few tests to fail. Don't merge yet.

**Comment by tjhei** [comment]
Hrm, I am wondering if those functions could atleast accept the block and the nonblock version of the dynamic sparsity pattern?

**Comment by drwells** [comment]
I fixed the tests (there were extra `<>`s).

@tjhei I do not understand what you mean. One function is designed for block dynamic sparsity patterns, and the other is designed for non-block dynamic sparsity patterns. What should we add?

**Comment by tjhei** [comment]
argh, sorry. Yes, for this function it is fine. I was just thinking that there are probably some other places in the library where the template is warranted because it could accept block and nonblock.

**Comment by drwells** [comment]
Yes, that is an even thornier issue than #1941 (I did leave `BlockSparsityPatternType` in one place where it really looked like it was wrong to call it with anything else). My only goal for this release is to make the current set of templates consistent and documented before adding more 'restrictions'.

File: source/lac/sparsity_tools.cc [hunk 1]
```diff
@@ -522,29 +522,31 @@ namespace SparsityTools
...
+   const std::vector<typename DynamicSparsityPattern::size_type> &rows_per_cpu,
+   const MPI_Comm                                                &mpi_comm,
+   const IndexSet                                                &myrange)
   {
     const unsigned int myid = Utilities::MPI::this_mpi_process(mpi_comm);
-    std::vector<typename DSP_t::size_type> start_index(rows_per_cpu.size()+1);
+    std::vector<typename DynamicSparsityPattern::size_type> start_index(rows_per_cpu.size()+1);
     start_index[0]=0;
-    for (typename DSP_t::size_type i=0; i<rows_per_cpu.size(); ++i)
+    for (typename DynamicSparsityPattern::size_type i=0; i<rows_per_cpu.size(); ++i)
```
Comment (bangerth) [inline source/lac/sparsity_tools.cc]: You don't need the `typename` here and it is in fact wrong in C++98. Same above and possibly in other places below.

File: include/deal.II/lac/sparsity_tools.h [hunk 2]
```diff
@@ -171,14 +172,14 @@ namespace SparsityTools
...
-   * Communicate rows in a compressed sparsity pattern over MPI.
+   * Communicate rows in a dynamic sparsity pattern over MPI.
    *
-   * @param dsp is the sparsity pattern that has been built locally and for
-   * which we need to exchange entries with other processors to make sure that
-   * each processor knows all the elements of the rows of a matrix it stores
-   * and that may eventually be written to. This sparsity pattern will be
-   * changed as a result of this function: All entries in rows that belong to
-   * a different processor are sent to them and added there.
+   * @param dsp is a dynamic sparsity pattern that has been built locally and
```
Comment (bangerth) [inline include/deal.II/lac/sparsity_tools.h]: The way it is rendered by doxygen, it looks better if the name of the argument is not part of a sentence. So write it as
```
  @param dsp A sparsity pattern that has been ...
```

**Comment by bangerth** [comment]
Please
- fix the typename issues
- the documentation part
- add a changelog entry
and merge yourself unless @tjhei has other comments as well.

File: source/lac/sparsity_tools.cc [hunk 3]
```diff
@@ -522,29 +522,31 @@ namespace SparsityTools
...
 
 #ifdef DEAL_II_WITH_MPI
-  template <class DSP_t>
-  void distribute_sparsity_pattern(DSP_t &dsp,
-                                   const std::vector<typename DSP_t::size_type> &rows_per_cpu,
-                                   const MPI_Comm &mpi_comm,
-                                   const IndexSet &myrange)
+  void distribute_sparsity_pattern
+  (DynamicSparsityPattern                                        &dsp,
+   const std::vector<typename DynamicSparsityPattern::size_type> &rows_per_cpu,
```
Comment (tjhei) [inline source/lac/sparsity_tools.cc]: the typename in the arguments also needs to go

**Pull request merged**